source/extensions/filters/network/dubbo_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,342 @@
+    conn_data_.reset();
+  }
+}
+
+void Router::UpstreamRequest::encodeData(Buffer::Instance& data) {
+  if (!conn_data_) {
+    ENVOY_STREAM_LOG(trace, "buffering {} bytes", *parent_.callbacks_, data.length());
+    if (!buffered_request_body_) {
+      buffered_request_body_ = std::make_unique<Envoy::Buffer::OwnedImpl>();
+    }
+
+    buffered_request_body_->move(data);
+  } else {
+    ENVOY_STREAM_LOG(trace, "proxying {} bytes", *parent_.callbacks_, data.length());
+    conn_data_->connection().write(data, false);
+  }
+}
+
+void Router::UpstreamRequest::onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                                            Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(warn, "dubbo upstream request: connection failure '{}'", host->address()->asString());
+
+  conn_pool_handle_ = nullptr;
+
+  // Mimic an upstream reset.
+  onUpstreamHostSelected(host);
+  onResetStream(reason);
+
+  parent_.upstream_request_buffer_.drain(parent_.upstream_request_buffer_.length());
+}
+
+void Router::UpstreamRequest::onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn_data,
+                                          Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(debug, "dubbo upstream request: tcp connection has ready");
+
+  // Only invoke continueDecoding if we'd previously stopped the filter chain.
+  bool continue_decoding = conn_pool_handle_ != nullptr;
+
+  onUpstreamHostSelected(host);
+  conn_data_ = std::move(conn_data);
+  conn_data_->addUpstreamCallbacks(parent_);
+  conn_pool_handle_ = nullptr;
+
+  onRequestStart(continue_decoding);
+}
+
+void Router::UpstreamRequest::onRequestStart(bool continue_decoding) {
+  ENVOY_LOG(debug, "dubbo upstream request: start sending data to the server {}",
+            upstream_host_->address()->asString());
+
+  if (buffered_request_body_) {
+    conn_data_->connection().write(*buffered_request_body_, false);
+    parent_.callbacks_->streamInfo().onLastUpstreamTxByteSent();
+  }
+
+  if (continue_decoding) {
+    parent_.callbacks_->continueDecoding();
+  }
+  onRequestComplete();
+}
+
+void Router::UpstreamRequest::onRequestComplete() { request_complete_ = true; }
+
+void Router::UpstreamRequest::onResponseComplete() {
+  response_complete_ = true;
+  conn_data_.reset();
+}
+
+void Router::UpstreamRequest::onUpstreamHostSelected(Upstream::HostDescriptionConstSharedPtr host) {
+  ENVOY_LOG(debug, "dubbo upstream request: selected upstream {}", host->address()->asString());
+  upstream_host_ = host;
+}
+
+void Router::UpstreamRequest::onResetStream(Tcp::ConnectionPool::PoolFailureReason reason) {
+  if (metadata_->message_type() == MessageType::Oneway) {
+    // For oneway requests, we should not attempt a response. Reset the downstream to signal
+    // an error.
+    ENVOY_LOG(debug, "dubbo upstream request: the request is oneway, reset downstream connection");
+    parent_.callbacks_->resetDownstreamConnection();
```

Comment:
Need to close the connection here?

Reply:
I change it to resetStream to end the current stream instead of closing the connection.